crates/matrix-sdk-ui/src/timeline/polls.rs

```diff
@@ -0,0 +1,182 @@
+        clone.response_data.push(ResponseData {
+            sender: sender.to_owned(),
+            timestamp,
+            answers: content.poll_response.answers.clone(),
+        });
+        clone
+    }
+
+    pub(super) fn end(&self, timestamp: MilliSecondsSinceUnixEpoch) -> Self {
+        let mut clone = self.clone();
+        clone.end_event_timestamp = Some(timestamp);
+        clone
+    }
+
+    pub fn fallback_text(&self) -> Option<String> {
+        self.start_event_content.text.clone()
+    }
+
+    pub fn results(&self) -> PollResult {
+        let results = compile_unstable_poll_results(
+            &self.start_event_content.poll_start,
+            self.response_data.iter().map(|response_data| PollResponseData {
+                sender: &response_data.sender,
+                origin_server_ts: response_data.timestamp,
+                selections: &response_data.answers,
+            }),
+            self.end_event_timestamp,
+        );
+
+        PollResult {
+            question: self.start_event_content.poll_start.question.text.clone(),
+            kind: self.start_event_content.poll_start.kind.clone(),
+            max_selections: self.start_event_content.poll_start.max_selections.into(),
+            answers: self
+                .start_event_content
+                .poll_start
+                .answers
+                .iter()
+                .map(|i| PollResultAnswer { id: i.id.clone(), text: i.text.clone() })
+                .collect(),
+            votes: results
+                .iter()
+                .map(|i| (i.0.to_string(), i.1.into_iter().map(|i| i.to_string()).collect()))
+                .collect(),
+            end_time: self.end_event_timestamp.map(|millis| millis.0.into()),
+        }
+    }
+}
+
+impl From<PollState> for UnstablePollStartEventContent {
+    fn from(value: PollState) -> Self {
+        let content = UnstablePollStartContentBlock::new(
+            value.start_event_content.poll_start.question.text.clone(),
+            value.start_event_content.poll_start.answers.clone(),
+        );
+        if let Some(text) = value.fallback_text() {
+            UnstablePollStartEventContent::plain_text(text, content)
+        } else {
+            UnstablePollStartEventContent::new(content)
+        }
+    }
+}
+
+/// Acts as a cache for poll response and poll end events handled before their
+/// start event has been handled.
+#[derive(Debug, Default)]
+pub(super) struct PollPendingEvents {
+    pub(super) pending_poll_responses: HashMap<OwnedEventId, Vec<ResponseData>>,
+    pub(super) pending_poll_ends: HashMap<OwnedEventId, MilliSecondsSinceUnixEpoch>,
+}
+
+impl PollPendingEvents {
+    pub(super) fn add_response(
+        &mut self,
+        start_id: &EventId,
+        sender: &UserId,
+        timestamp: MilliSecondsSinceUnixEpoch,
+        content: &UnstablePollResponseEventContent,
+    ) {
+        self.pending_poll_responses.entry(start_id.to_owned()).or_insert(vec![]).push(
```

Comment:
```suggestion
        self.pending_poll_responses.entry(start_id.to_owned()).or_default().push(
```

Reply:
Is there any difference besides style?

Reply:
There is a tiny difference in that before, the vector was instantiated regardless of whether the entry was found or not. However, since creating an empty vector doesn't allocate, it should not make an actual difference in runtime code, at least for release builds.